core/storage/trie/polkadot_trie/polkadot_trie_cursor_impl.cpp

```diff
@@ -101,159 +110,149 @@ namespace kagome::storage::trie {
           }
-          followed_path.emplace_back(current_as_branch,
-                                     left_nibbles[common_length]);
-          current = n;
-          left_nibbles = left_nibbles.subspan(common_length + 1);
-          break;
+          case NodeType::Leaf: {
+            break;  // go to case3
+          }
+          default:
+            return Error::INVALID_NODE_TYPE;
         }
-        case NodeType::Leaf:
-          left_nibbles = left_nibbles.subspan(common_length);
-          goto END;
-
-        case NodeType::Special:
-          return Error::INVALID_NODE_TYPE;
+      }
+      // if the left part of the key is longer than the current of
+      // lexicographically greater than the current, we must return to its
+      // parent and find a child greater than the current one
+      bool longerOrBigger =
+          longer or (not part_equal and *left_mismatch > *current_mismatch);
+      if (longerOrBigger) {
+        while (not last_visited_child_.empty()) {
+          auto [parent, idx] = last_visited_child_.back();
+          last_visited_child_.pop_back();
+          OUTCOME_TRY(child_idx, getChildWithMinIdx(parent, idx + 1));
+          if (child_idx != -1) {
+            OUTCOME_TRY(child, trie_.retrieveChild(parent, child_idx));
+            OUTCOME_TRY(node, seekNodeWithValue(child));
+            current_ = node;
+            return outcome::success();
+          }
+        }
+        current_ = nullptr;
+        return outcome::success();
       }
     }
-  END:
-    return followed_path;
+    BOOST_ASSERT_MSG(false, "Unreachable");
   }
 
-  outcome::result<boost::optional<PolkadotTrieCursorImpl::NodePtr>>
-  PolkadotTrieCursorImpl::seekChildWithValue(BranchPtr node) {
-    return seekChildWithValueAfterIdx(node, -1);
-  }
-
-  outcome::result<boost::optional<PolkadotTrieCursorImpl::NodePtr>>
-  PolkadotTrieCursorImpl::seekChildWithValueAfterIdx(BranchPtr node, int8_t idx) {
-    if (node == nullptr) {
-      return boost::none;
-    }
-    do {
-      for (uint8_t i = idx + 1; i < BranchNode::kMaxChildren; i++) {
-        OUTCOME_TRY(child, trie_.retrieveChild(node, i));
+  outcome::result<PolkadotTrieCursorImpl::NodePtr>
+  PolkadotTrieCursorImpl::seekNodeWithValue(NodePtr node) {
+    while (not node->value.has_value()) {
+      if (not node->isBranch()) {
+        return Error::INVALID_NODE_TYPE;  // can't be a leaf without a value
+      }
+      auto node_as_value = std::dynamic_pointer_cast<BranchNode>(node);
+      for (uint8_t i = 0; i < BranchNode::kMaxChildren; i++) {
+        OUTCOME_TRY(child, trie_.retrieveChild(node_as_value, i));
         if (child != nullptr) {
-          last_visited_child_.emplace_back(node, i);
+          last_visited_child_.emplace_back(node_as_value, i);
           switch (child->getTrieType()) {
             case NodeType::BranchEmptyValue:
               node = std::dynamic_pointer_cast<BranchNode>(child);
-              break;
+              goto BREAK;
```

Comment:
How I know clang-tidy doesn't like **goto** keyword =)

Reply:
I guess, we can ust to return node here

Reply:
Nope, goto is used to break out of inner for to outer while. Clang-tidy didn't object on CI, so I hope it understands that in this case GOTO is totally justified.